{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d061f73f-f0f0-4582-9f4d-93aa4a847bc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Choosing genes for analysis ###\n",
    "\n",
    "### Citations of tools ###\n",
    "# gget: Luebbert, L., & Pachter, L. (2023). Efficient querying of genomic reference databases with gget. Bioinformatics. https://doi.org/10.1093/bioinformatics/btac836\n",
    "# pybiomart: https://github.com/jrderuiter/pybiomart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b020c68d-137c-4b0a-ae2b-9a6db7450650",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import gget\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "4189d992-1ff8-41b2-9580-1c6aa9ee166a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data file downloaded from https://www.encodeproject.org/experiments/ENCSR792OIJ/\n",
    "# Accession ENCFF928NYA under processed data (tsv file) \n",
    "# Data added 2020-11-04\n",
    "df = pd.read_table(\"ENCFF928NYA.tsv\")\n",
    "columns = ['gene_id', 'TPM', 'FPKM', 'transcript_id(s)']\n",
    "df = pd.DataFrame(df, columns=columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "2da92b03-41ba-4c84-8c64-89e972489124",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>transcript_id(s)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10904</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12954</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12956</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12958</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12960</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12960</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id  TPM  FPKM transcript_id(s)\n",
       "0   10904  0.0   0.0            10904\n",
       "1   12954  0.0   0.0            12954\n",
       "2   12956  0.0   0.0            12956\n",
       "3   12958  0.0   0.0            12958\n",
       "4   12960  0.0   0.0            12960"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b7f10d6e-ad37-4e81-978f-829bfd0cbb7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "expressed_df = df.query(\"TPM!=0.0\")\n",
    "\n",
    "high_thresh = 200\n",
    "high_thresh_df = expressed_df.query(\"TPM>@high_thresh\")\n",
    "\n",
    "low_thresh = 10\n",
    "low_thresh_df = expressed_df.query(\"TPM>5 & TPM<@low_thresh\")\n",
    "lowest_thresh_df = expressed_df.query(\"TPM<5\")\n",
    "\n",
    "# mid_thresh_df = high_thresh_df.query(\"TPM<20000\")\n",
    "# mid_thresh_df = mid_thresh_df.drop(index=[59429, 59431, 59456, 59471, 59484, 59493, 59500, 59524])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b9130faf-ba43-4680-9180-e7fcd01e1026",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(high_thresh_df.to_string())\n",
    "#print(len(mid_thresh_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "97db157c-ed5c-45c5-b250-ba93428c2aaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>13023</td>\n",
       "      <td>41.07</td>\n",
       "      <td>49.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>13024</td>\n",
       "      <td>16.31</td>\n",
       "      <td>19.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>30958</td>\n",
       "      <td>43.49</td>\n",
       "      <td>52.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471</th>\n",
       "      <td>30964</td>\n",
       "      <td>52.81</td>\n",
       "      <td>63.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_id    TPM   FPKM\n",
       "46                13023  41.07  49.61\n",
       "47                13024  16.31  19.70\n",
       "465               30958  43.49  52.53\n",
       "471               30964  52.81  63.79\n",
       "649  ENSG00000000003.14   0.05   0.06"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expressed_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "52f533ef-ba01-4c2d-a155-beaa77db2d5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rows were selected by looking at the entire dataframe and were chosen for varying expression levels\n",
    "high_rows = [941, 1965, 4142, 10644, 17808, 24929, 48356, 53485, 54599, 55026, 55115, 57853]\n",
    "low_rows = [657, 926]\n",
    "lowest_rows = [649, 652]\n",
    "\n",
    "chosen_rows = high_rows + low_rows + lowest_rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "333b81a4-0dba-4157-ab5f-55e7d0584c4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>ENSG00000011304.19</td>\n",
       "      <td>342.64</td>\n",
       "      <td>413.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1965</th>\n",
       "      <td>ENSG00000075624.14</td>\n",
       "      <td>1567.70</td>\n",
       "      <td>1893.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4142</th>\n",
       "      <td>ENSG00000108107.14</td>\n",
       "      <td>1401.16</td>\n",
       "      <td>1692.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10644</th>\n",
       "      <td>ENSG00000156508.17</td>\n",
       "      <td>5000.06</td>\n",
       "      <td>6039.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17808</th>\n",
       "      <td>ENSG00000196565.14</td>\n",
       "      <td>4169.72</td>\n",
       "      <td>5036.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24929</th>\n",
       "      <td>ENSG00000222328.1</td>\n",
       "      <td>7046.27</td>\n",
       "      <td>8511.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48356</th>\n",
       "      <td>ENSG00000263934.4</td>\n",
       "      <td>13928.59</td>\n",
       "      <td>16824.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53485</th>\n",
       "      <td>ENSG00000274012.1</td>\n",
       "      <td>94300.79</td>\n",
       "      <td>113903.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54599</th>\n",
       "      <td>ENSG00000276168.1</td>\n",
       "      <td>149831.97</td>\n",
       "      <td>180978.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55026</th>\n",
       "      <td>ENSG00000277027.1</td>\n",
       "      <td>23024.05</td>\n",
       "      <td>27810.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55115</th>\n",
       "      <td>ENSG00000277209.1</td>\n",
       "      <td>43982.92</td>\n",
       "      <td>53125.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57853</th>\n",
       "      <td>ENSG00000283293.1</td>\n",
       "      <td>66929.44</td>\n",
       "      <td>80842.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657</th>\n",
       "      <td>ENSG00000001084.12</td>\n",
       "      <td>6.26</td>\n",
       "      <td>7.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926</th>\n",
       "      <td>ENSG00000011021.22</td>\n",
       "      <td>9.82</td>\n",
       "      <td>11.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>ENSG00000000457.13</td>\n",
       "      <td>3.32</td>\n",
       "      <td>4.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_id        TPM       FPKM\n",
       "941    ENSG00000011304.19     342.64     413.86\n",
       "1965   ENSG00000075624.14    1567.70    1893.59\n",
       "4142   ENSG00000108107.14    1401.16    1692.44\n",
       "10644  ENSG00000156508.17    5000.06    6039.46\n",
       "17808  ENSG00000196565.14    4169.72    5036.51\n",
       "24929   ENSG00000222328.1    7046.27    8511.03\n",
       "48356   ENSG00000263934.4   13928.59   16824.04\n",
       "53485   ENSG00000274012.1   94300.79  113903.82\n",
       "54599   ENSG00000276168.1  149831.97  180978.68\n",
       "55026   ENSG00000277027.1   23024.05   27810.24\n",
       "55115   ENSG00000277209.1   43982.92   53125.98\n",
       "57853   ENSG00000283293.1   66929.44   80842.58\n",
       "657    ENSG00000001084.12       6.26       7.56\n",
       "926    ENSG00000011021.22       9.82      11.86\n",
       "649    ENSG00000000003.14       0.05       0.06\n",
       "652    ENSG00000000457.13       3.32       4.01"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chosen_rows_df = pd.DataFrame(df.loc[chosen_rows])\n",
    "\n",
    "chosen_rows_gene_ids = []\n",
    "for index, row in chosen_rows_df.iterrows():\n",
    "    chosen_rows_gene_ids.append(row['gene_id'])\n",
    "    \n",
    "chosen_rows_df.head(n=len(chosen_rows_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e365a725-64c3-4883-afc6-21d00abd6211",
   "metadata": {},
   "outputs": [],
   "source": [
    "### METHOD USING GGET ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "db428d50-2607-444d-a4e9-7d40453b62e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Jun 30 03:35:01 2023 INFO We noticed that you passed a version number with your Ensembl ID.\n",
      "Please note that gget info will always return information linked to the latest Ensembl ID version (see 'ensembl_id').\n"
     ]
    }
   ],
   "source": [
    "res = gget.info(chosen_rows_gene_ids, json=True, uniprot=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "13274da2-9161-48ca-ab97-4c46fd657101",
   "metadata": {},
   "outputs": [],
   "source": [
    "ordered_gene_names = [None] * len(chosen_rows_gene_ids)\n",
    "\n",
    "for key in res.keys():\n",
    "    for i in range(len(chosen_rows_gene_ids)):\n",
    "        if chosen_rows_gene_ids[i].startswith(key):\n",
    "            ordered_gene_names[i] = res[key]['ensembl_gene_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7ac14023-a8ad-4058-8bc2-917d011e83c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene name (ensembl)</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>ENSG00000011304.19</td>\n",
       "      <td>PTBP1</td>\n",
       "      <td>342.64</td>\n",
       "      <td>413.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1965</th>\n",
       "      <td>ENSG00000075624.14</td>\n",
       "      <td>ACTB</td>\n",
       "      <td>1567.70</td>\n",
       "      <td>1893.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4142</th>\n",
       "      <td>ENSG00000108107.14</td>\n",
       "      <td>RPL28</td>\n",
       "      <td>1401.16</td>\n",
       "      <td>1692.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10644</th>\n",
       "      <td>ENSG00000156508.17</td>\n",
       "      <td>EEF1A1</td>\n",
       "      <td>5000.06</td>\n",
       "      <td>6039.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17808</th>\n",
       "      <td>ENSG00000196565.14</td>\n",
       "      <td>HBG2</td>\n",
       "      <td>4169.72</td>\n",
       "      <td>5036.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24929</th>\n",
       "      <td>ENSG00000222328.1</td>\n",
       "      <td>RNU2-2P</td>\n",
       "      <td>7046.27</td>\n",
       "      <td>8511.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48356</th>\n",
       "      <td>ENSG00000263934.4</td>\n",
       "      <td>SNORD3A</td>\n",
       "      <td>13928.59</td>\n",
       "      <td>16824.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53485</th>\n",
       "      <td>ENSG00000274012.1</td>\n",
       "      <td>RN7SL2</td>\n",
       "      <td>94300.79</td>\n",
       "      <td>113903.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54599</th>\n",
       "      <td>ENSG00000276168.1</td>\n",
       "      <td>RN7SL1</td>\n",
       "      <td>149831.97</td>\n",
       "      <td>180978.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55026</th>\n",
       "      <td>ENSG00000277027.1</td>\n",
       "      <td>RMRP</td>\n",
       "      <td>23024.05</td>\n",
       "      <td>27810.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55115</th>\n",
       "      <td>ENSG00000277209.1</td>\n",
       "      <td>RPPH1</td>\n",
       "      <td>43982.92</td>\n",
       "      <td>53125.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57853</th>\n",
       "      <td>ENSG00000283293.1</td>\n",
       "      <td>RN7SK</td>\n",
       "      <td>66929.44</td>\n",
       "      <td>80842.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657</th>\n",
       "      <td>ENSG00000001084.12</td>\n",
       "      <td>GCLC</td>\n",
       "      <td>6.26</td>\n",
       "      <td>7.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926</th>\n",
       "      <td>ENSG00000011021.22</td>\n",
       "      <td>CLCN6</td>\n",
       "      <td>9.82</td>\n",
       "      <td>11.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>ENSG00000000457.13</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>3.32</td>\n",
       "      <td>4.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_id gene name (ensembl)        TPM       FPKM\n",
       "941    ENSG00000011304.19               PTBP1     342.64     413.86\n",
       "1965   ENSG00000075624.14                ACTB    1567.70    1893.59\n",
       "4142   ENSG00000108107.14               RPL28    1401.16    1692.44\n",
       "10644  ENSG00000156508.17              EEF1A1    5000.06    6039.46\n",
       "17808  ENSG00000196565.14                HBG2    4169.72    5036.51\n",
       "24929   ENSG00000222328.1             RNU2-2P    7046.27    8511.03\n",
       "48356   ENSG00000263934.4             SNORD3A   13928.59   16824.04\n",
       "53485   ENSG00000274012.1              RN7SL2   94300.79  113903.82\n",
       "54599   ENSG00000276168.1              RN7SL1  149831.97  180978.68\n",
       "55026   ENSG00000277027.1                RMRP   23024.05   27810.24\n",
       "55115   ENSG00000277209.1               RPPH1   43982.92   53125.98\n",
       "57853   ENSG00000283293.1               RN7SK   66929.44   80842.58\n",
       "657    ENSG00000001084.12                GCLC       6.26       7.56\n",
       "926    ENSG00000011021.22               CLCN6       9.82      11.86\n",
       "649    ENSG00000000003.14              TSPAN6       0.05       0.06\n",
       "652    ENSG00000000457.13               SCYL3       3.32       4.01"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chosen_rows_df.insert(1, \"gene name (ensembl)\", ordered_gene_names, True)\n",
    "chosen_rows_df.head(n=len(chosen_rows_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6f55cf83-80cd-45ac-bc78-cc4e92fb9897",
   "metadata": {},
   "outputs": [],
   "source": [
    "# However, we need the RefSeq ID beginning in NM_ to use PaintSHOP, and gget doesn't let us do that. \n",
    "# So, the following method uses pybiomart which uses the ensembl API and MANE (Matched Annotation from the NCBI and EMBL-EBI) which correlates RefSeq IDs to Ensembl IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d50ead79-9d54-4f09-83a5-efd435f319e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pybiomart import Dataset\n",
    "\n",
    "dataset = Dataset(name='hsapiens_gene_ensembl',\n",
    "                  host='http://www.ensembl.org')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "75c00436-8cd8-4b94-a898-b1a6dcade9a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "refseq_results = []\n",
    "\n",
    "# Gets the gene name and the RefSeq ID associated with the canonical (most common) transcript\n",
    "for i in chosen_rows_gene_ids:\n",
    "    gene_id = i.split(\".\")[0] # remove the version number\n",
    "    result = dataset.query(attributes=['ensembl_gene_id', 'external_gene_name', 'external_transcript_name', 'transcript_mane_select', 'transcript_is_canonical', 'ensembl_transcript_id'],\n",
    "                  filters={'link_ensembl_gene_id': [gene_id], 'transcript_is_canonical' : True})\n",
    "    refseq_results.append(result)\n",
    "    \n",
    "to_extract = chosen_rows_df.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "603f2fb5-fe6f-4260-adbc-f0fdf3060772",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Gene stable ID Gene name        TPM       FPKM Transcript name RefSeq match transcript (MANE Select)  Ensembl Canonical Transcript stable ID\n",
      "0   ENSG00000011304     PTBP1     342.64     413.86       PTBP1-203                           NM_002819.5               True      ENST00000356948\n",
      "1   ENSG00000075624      ACTB    1567.70    1893.59        ACTB-217                           NM_001101.5               True      ENST00000646664\n",
      "2   ENSG00000108107     RPL28    1401.16    1692.44       RPL28-201                           NM_000991.5               True      ENST00000344063\n",
      "3   ENSG00000156508    EEF1A1    5000.06    6039.46      EEF1A1-201                           NM_001402.6               True      ENST00000309268\n",
      "4   ENSG00000196565      HBG2    4169.72    5036.51        HBG2-201                           NM_000184.3               True      ENST00000336906\n",
      "5   ENSG00000222328   RNU2-2P    7046.27    8511.03     RNU2-2P-201                                   NaN               True      ENST00000410396\n",
      "6   ENSG00000263934   SNORD3A   13928.59   16824.04     SNORD3A-201                                   NaN               True      ENST00000584923\n",
      "7   ENSG00000274012    RN7SL2   94300.79  113903.82      RN7SL2-201                                   NaN               True      ENST00000490232\n",
      "8   ENSG00000276168    RN7SL1  149831.97  180978.68      RN7SL1-201                                   NaN               True      ENST00000618786\n",
      "9   ENSG00000277027      RMRP   23024.05   27810.24        RMRP-201                                   NaN               True      ENST00000363046\n",
      "10  ENSG00000277209     RPPH1   43982.92   53125.98       RPPH1-201                                   NaN               True      ENST00000516869\n",
      "11  ENSG00000283293     RN7SK   66929.44   80842.58       RN7SK-201                                   NaN               True      ENST00000636484\n",
      "12  ENSG00000001084      GCLC       6.26       7.56        GCLC-214                           NM_001498.4               True      ENST00000650454\n",
      "13  ENSG00000011021     CLCN6       9.82      11.86       CLCN6-202                           NM_001286.5               True      ENST00000346436\n",
      "14  ENSG00000000003    TSPAN6       0.05       0.06      TSPAN6-201                           NM_003270.4               True      ENST00000373020\n",
      "15  ENSG00000000457     SCYL3       3.32       4.01       SCYL3-202                           NM_020423.7               True      ENST00000367771\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>Transcript name</th>\n",
       "      <th>RefSeq match transcript (MANE Select)</th>\n",
       "      <th>Ensembl Canonical</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000011304</td>\n",
       "      <td>PTBP1</td>\n",
       "      <td>342.64</td>\n",
       "      <td>413.86</td>\n",
       "      <td>PTBP1-203</td>\n",
       "      <td>NM_002819.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000356948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000075624</td>\n",
       "      <td>ACTB</td>\n",
       "      <td>1567.70</td>\n",
       "      <td>1893.59</td>\n",
       "      <td>ACTB-217</td>\n",
       "      <td>NM_001101.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000646664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000108107</td>\n",
       "      <td>RPL28</td>\n",
       "      <td>1401.16</td>\n",
       "      <td>1692.44</td>\n",
       "      <td>RPL28-201</td>\n",
       "      <td>NM_000991.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000344063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000156508</td>\n",
       "      <td>EEF1A1</td>\n",
       "      <td>5000.06</td>\n",
       "      <td>6039.46</td>\n",
       "      <td>EEF1A1-201</td>\n",
       "      <td>NM_001402.6</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000309268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000196565</td>\n",
       "      <td>HBG2</td>\n",
       "      <td>4169.72</td>\n",
       "      <td>5036.51</td>\n",
       "      <td>HBG2-201</td>\n",
       "      <td>NM_000184.3</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000336906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000222328</td>\n",
       "      <td>RNU2-2P</td>\n",
       "      <td>7046.27</td>\n",
       "      <td>8511.03</td>\n",
       "      <td>RNU2-2P-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000410396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000263934</td>\n",
       "      <td>SNORD3A</td>\n",
       "      <td>13928.59</td>\n",
       "      <td>16824.04</td>\n",
       "      <td>SNORD3A-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000584923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000274012</td>\n",
       "      <td>RN7SL2</td>\n",
       "      <td>94300.79</td>\n",
       "      <td>113903.82</td>\n",
       "      <td>RN7SL2-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000490232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000276168</td>\n",
       "      <td>RN7SL1</td>\n",
       "      <td>149831.97</td>\n",
       "      <td>180978.68</td>\n",
       "      <td>RN7SL1-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000618786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000277027</td>\n",
       "      <td>RMRP</td>\n",
       "      <td>23024.05</td>\n",
       "      <td>27810.24</td>\n",
       "      <td>RMRP-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000363046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000277209</td>\n",
       "      <td>RPPH1</td>\n",
       "      <td>43982.92</td>\n",
       "      <td>53125.98</td>\n",
       "      <td>RPPH1-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000516869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000283293</td>\n",
       "      <td>RN7SK</td>\n",
       "      <td>66929.44</td>\n",
       "      <td>80842.58</td>\n",
       "      <td>RN7SK-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000636484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000001084</td>\n",
       "      <td>GCLC</td>\n",
       "      <td>6.26</td>\n",
       "      <td>7.56</td>\n",
       "      <td>GCLC-214</td>\n",
       "      <td>NM_001498.4</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000650454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000011021</td>\n",
       "      <td>CLCN6</td>\n",
       "      <td>9.82</td>\n",
       "      <td>11.86</td>\n",
       "      <td>CLCN6-202</td>\n",
       "      <td>NM_001286.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000346436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.06</td>\n",
       "      <td>TSPAN6-201</td>\n",
       "      <td>NM_003270.4</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000373020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>3.32</td>\n",
       "      <td>4.01</td>\n",
       "      <td>SCYL3-202</td>\n",
       "      <td>NM_020423.7</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000367771</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene stable ID Gene name        TPM       FPKM Transcript name  \\\n",
       "0   ENSG00000011304     PTBP1     342.64     413.86       PTBP1-203   \n",
       "1   ENSG00000075624      ACTB    1567.70    1893.59        ACTB-217   \n",
       "2   ENSG00000108107     RPL28    1401.16    1692.44       RPL28-201   \n",
       "3   ENSG00000156508    EEF1A1    5000.06    6039.46      EEF1A1-201   \n",
       "4   ENSG00000196565      HBG2    4169.72    5036.51        HBG2-201   \n",
       "5   ENSG00000222328   RNU2-2P    7046.27    8511.03     RNU2-2P-201   \n",
       "6   ENSG00000263934   SNORD3A   13928.59   16824.04     SNORD3A-201   \n",
       "7   ENSG00000274012    RN7SL2   94300.79  113903.82      RN7SL2-201   \n",
       "8   ENSG00000276168    RN7SL1  149831.97  180978.68      RN7SL1-201   \n",
       "9   ENSG00000277027      RMRP   23024.05   27810.24        RMRP-201   \n",
       "10  ENSG00000277209     RPPH1   43982.92   53125.98       RPPH1-201   \n",
       "11  ENSG00000283293     RN7SK   66929.44   80842.58       RN7SK-201   \n",
       "12  ENSG00000001084      GCLC       6.26       7.56        GCLC-214   \n",
       "13  ENSG00000011021     CLCN6       9.82      11.86       CLCN6-202   \n",
       "14  ENSG00000000003    TSPAN6       0.05       0.06      TSPAN6-201   \n",
       "15  ENSG00000000457     SCYL3       3.32       4.01       SCYL3-202   \n",
       "\n",
       "   RefSeq match transcript (MANE Select)  Ensembl Canonical  \\\n",
       "0                            NM_002819.5               True   \n",
       "1                            NM_001101.5               True   \n",
       "2                            NM_000991.5               True   \n",
       "3                            NM_001402.6               True   \n",
       "4                            NM_000184.3               True   \n",
       "5                                    NaN               True   \n",
       "6                                    NaN               True   \n",
       "7                                    NaN               True   \n",
       "8                                    NaN               True   \n",
       "9                                    NaN               True   \n",
       "10                                   NaN               True   \n",
       "11                                   NaN               True   \n",
       "12                           NM_001498.4               True   \n",
       "13                           NM_001286.5               True   \n",
       "14                           NM_003270.4               True   \n",
       "15                           NM_020423.7               True   \n",
       "\n",
       "   Transcript stable ID  \n",
       "0       ENST00000356948  \n",
       "1       ENST00000646664  \n",
       "2       ENST00000344063  \n",
       "3       ENST00000309268  \n",
       "4       ENST00000336906  \n",
       "5       ENST00000410396  \n",
       "6       ENST00000584923  \n",
       "7       ENST00000490232  \n",
       "8       ENST00000618786  \n",
       "9       ENST00000363046  \n",
       "10      ENST00000516869  \n",
       "11      ENST00000636484  \n",
       "12      ENST00000650454  \n",
       "13      ENST00000346436  \n",
       "14      ENST00000373020  \n",
       "15      ENST00000367771  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_df = pd.concat(refseq_results, ignore_index = True)\n",
    "refseq_df.insert(2, \"TPM\", to_extract[\"TPM\"])\n",
    "refseq_df.insert(3, \"FPKM\", to_extract[\"FPKM\"])\n",
    "refseq_df['Ensembl Canonical'] = refseq_df['Ensembl Canonical'].astype(bool)  \n",
    "print(refseq_df.to_string())\n",
    "refseq_df.head(n=len(refseq_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f34050eb-8a73-43ab-8efd-7a08f248d4d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mergedRes = pd.merge(expressed_df, refseq_df, on ='gene_id')\n",
    "# mergedRes.head(len(mergedRes.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "96642316-9280-4da2-810c-91ea0d1717a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note that several of the genes have no matching RefSeq ID. This is because they are not mRNA, but other kinds of RNA (i.e. snRNA or rRNA). \n",
    "# Ensembl doesn't already have the RefSeq IDs available for this. --> see PartialAnalysis notebook to filter these out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "810ed68f-256b-4872-b5b6-6881b50ffd9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "valid_seqs = []\n",
    "# Export to file to use in PaintSHOP\n",
    "f = open(\"refseqs_sample.txt\", \"w\")\n",
    "for index, row in refseq_df.iterrows():\n",
    "    if str(row['RefSeq match transcript (MANE Select)']) != 'nan':\n",
    "        valid_seqs.append(str(row['RefSeq match transcript (MANE Select)']))\n",
    "        f.write(str(row['RefSeq match transcript (MANE Select)']) + '\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "08adaab0-cb19-4bc7-96c8-2d5191004956",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Then we use PaintSHOP to make probes: https://doi.org/10.1038/s41592-021-01187-3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "b66f6a1b-786a-44e3-a905-51407ebb20c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_table('230629-PaintSHOP-probes.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "299a9000-311c-4fda-ba7b-29f0c326f7db",
   "metadata": {},
   "outputs": [],
   "source": [
    "#data.head(len(data.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "fc1da3c6-8c80-4c2e-9405-ff7fb81faeea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>sequence</th>\n",
       "      <th>Tm</th>\n",
       "      <th>on_target</th>\n",
       "      <th>off_target</th>\n",
       "      <th>repeat_seq</th>\n",
       "      <th>prob</th>\n",
       "      <th>max_kmer</th>\n",
       "      <th>probe_strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NM_001101</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5528340</td>\n",
       "      <td>5528369</td>\n",
       "      <td>ATGACCTGGCCGTCAGGCAGCTCGTAGCTC</td>\n",
       "      <td>46.30</td>\n",
       "      <td>100.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.064</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_001101</td>\n",
       "      <td>ACTB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NM_001101</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5528658</td>\n",
       "      <td>5528688</td>\n",
       "      <td>AGGGATAGCACAGCCTGGATAGCAACGTACA</td>\n",
       "      <td>42.32</td>\n",
       "      <td>100.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.277</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_001101</td>\n",
       "      <td>ACTB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NM_001101</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5529286</td>\n",
       "      <td>5529315</td>\n",
       "      <td>CCCAGTTGGTGACGATGCCGTGCTCGATGG</td>\n",
       "      <td>45.84</td>\n",
       "      <td>99.194</td>\n",
       "      <td>99.194</td>\n",
       "      <td>0</td>\n",
       "      <td>0.226</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_001101</td>\n",
       "      <td>ACTB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NM_001101</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5528567</td>\n",
       "      <td>5528596</td>\n",
       "      <td>GGGGAGGGCATACCCCTCGTAGATGGGCAC</td>\n",
       "      <td>46.15</td>\n",
       "      <td>97.338</td>\n",
       "      <td>100.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.051</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_001101</td>\n",
       "      <td>ACTB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NM_001101</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5528310</td>\n",
       "      <td>5528339</td>\n",
       "      <td>GGGCAGCGGAACCGCTCATTGCCAATGGTG</td>\n",
       "      <td>46.93</td>\n",
       "      <td>97.920</td>\n",
       "      <td>166.778</td>\n",
       "      <td>0</td>\n",
       "      <td>0.083</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_001101</td>\n",
       "      <td>ACTB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>NM_003270</td>\n",
       "      <td>chrX</td>\n",
       "      <td>100635578</td>\n",
       "      <td>100635608</td>\n",
       "      <td>AAGCTCGGCAGGTAGCAAAACAACCAAAGGT</td>\n",
       "      <td>42.60</td>\n",
       "      <td>100.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.239</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_003270</td>\n",
       "      <td>TSPAN6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>NM_003270</td>\n",
       "      <td>chrX</td>\n",
       "      <td>100636686</td>\n",
       "      <td>100636715</td>\n",
       "      <td>GGACGCCATGACTAGCCCGAGACCCTGCAC</td>\n",
       "      <td>46.66</td>\n",
       "      <td>98.682</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.351</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_003270</td>\n",
       "      <td>TSPAN6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>NM_003270</td>\n",
       "      <td>chrX</td>\n",
       "      <td>100635715</td>\n",
       "      <td>100635745</td>\n",
       "      <td>ATGCCAACTGCAAGAAGGATAACGCCAGTGA</td>\n",
       "      <td>42.11</td>\n",
       "      <td>100.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.168</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_003270</td>\n",
       "      <td>TSPAN6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>NM_003270</td>\n",
       "      <td>chrX</td>\n",
       "      <td>100627400</td>\n",
       "      <td>100627429</td>\n",
       "      <td>CCCCAGCCTGTGCCACTAAGGGAAACTCAC</td>\n",
       "      <td>43.90</td>\n",
       "      <td>100.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.181</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_003270</td>\n",
       "      <td>TSPAN6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>NM_003270</td>\n",
       "      <td>chrX</td>\n",
       "      <td>100627263</td>\n",
       "      <td>100627292</td>\n",
       "      <td>AACCTGGTGCCTGGAAATAAGGCAACCCAC</td>\n",
       "      <td>42.71</td>\n",
       "      <td>97.888</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.146</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_003270</td>\n",
       "      <td>TSPAN6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>147 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        refseq chrom      start       stop                         sequence  \\\n",
       "14   NM_001101  chr7    5528340    5528369   ATGACCTGGCCGTCAGGCAGCTCGTAGCTC   \n",
       "15   NM_001101  chr7    5528658    5528688  AGGGATAGCACAGCCTGGATAGCAACGTACA   \n",
       "18   NM_001101  chr7    5529286    5529315   CCCAGTTGGTGACGATGCCGTGCTCGATGG   \n",
       "17   NM_001101  chr7    5528567    5528596   GGGGAGGGCATACCCCTCGTAGATGGGCAC   \n",
       "13   NM_001101  chr7    5528310    5528339   GGGCAGCGGAACCGCTCATTGCCAATGGTG   \n",
       "..         ...   ...        ...        ...                              ...   \n",
       "248  NM_003270  chrX  100635578  100635608  AAGCTCGGCAGGTAGCAAAACAACCAAAGGT   \n",
       "249  NM_003270  chrX  100636686  100636715   GGACGCCATGACTAGCCCGAGACCCTGCAC   \n",
       "250  NM_003270  chrX  100635715  100635745  ATGCCAACTGCAAGAAGGATAACGCCAGTGA   \n",
       "255  NM_003270  chrX  100627400  100627429   CCCCAGCCTGTGCCACTAAGGGAAACTCAC   \n",
       "257  NM_003270  chrX  100627263  100627292   AACCTGGTGCCTGGAAATAAGGCAACCCAC   \n",
       "\n",
       "        Tm  on_target  off_target  repeat_seq   prob  max_kmer probe_strand  \\\n",
       "14   46.30    100.000       0.000           0  0.064         2            +   \n",
       "15   42.32    100.000       0.000           0  0.277         2            +   \n",
       "18   45.84     99.194      99.194           0  0.226         3            +   \n",
       "17   46.15     97.338     100.000           0  0.051         2            +   \n",
       "13   46.93     97.920     166.778           0  0.083         2            +   \n",
       "..     ...        ...         ...         ...    ...       ...          ...   \n",
       "248  42.60    100.000       0.000           0  0.239         0            +   \n",
       "249  46.66     98.682       0.000           0  0.351         0            +   \n",
       "250  42.11    100.000       0.000           0  0.168         2            +   \n",
       "255  43.90    100.000       0.000           0  0.181         2            +   \n",
       "257  42.71     97.888       0.000           0  0.146         2            +   \n",
       "\n",
       "    transcript_id gene_id  \n",
       "14      NM_001101    ACTB  \n",
       "15      NM_001101    ACTB  \n",
       "18      NM_001101    ACTB  \n",
       "17      NM_001101    ACTB  \n",
       "13      NM_001101    ACTB  \n",
       "..            ...     ...  \n",
       "248     NM_003270  TSPAN6  \n",
       "249     NM_003270  TSPAN6  \n",
       "250     NM_003270  TSPAN6  \n",
       "255     NM_003270  TSPAN6  \n",
       "257     NM_003270  TSPAN6  \n",
       "\n",
       "[147 rows x 14 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mask = (data['sequence'].str.len() <= 31)\n",
    "data = data.loc[mask]\n",
    "data = data.sort_values(['gene_id', 'off_target'], ascending=[True, True])\n",
    "data.head(n=300)\n",
    "\n",
    "#data.query(\"'gene_id'=='EEF1A1'\").head(n=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "8a0d5354-7d16-4b3a-ab3a-e759161021a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NM_002819.5 40\n",
      "NM_001101.5 6\n",
      "NM_000991.5 0\n",
      "NM_001402.6 0\n",
      "NM_000184.3 6\n",
      "NM_001498.4 10\n",
      "NM_001286.5 61\n",
      "NM_003270.4 6\n",
      "NM_020423.7 18\n"
     ]
    }
   ],
   "source": [
    "gene_probes_dict = {}\n",
    "\n",
    "for i in valid_seqs:\n",
    "    gene = data.query(\"refseq=='\" + i.split(\".\")[0] + \"'\")\n",
    "    gene_probes_dict[i] = gene\n",
    "    print(i, len(gene.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "1110978c-0012-495c-ac35-12e5aadc5f27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# HGB1: \n",
    "# 2: one for HGB1 and another for HGB2 (one off target on ch2 --> not a gene)\n",
    "# 6: one for HGB1 and another for HGB2\n",
    "# 12: one for HGB1 and another for HGB2\n",
    "# 0: one for HGB1 and another for HGB2 (off target gene on ch6)\n",
    "# 12, 6, 2, 7, 8 (7, and 8 made smaller to fit)\n",
    "\n",
    "# ACTB:\n",
    "# 14: ch7 -> match to ACTB, ch6 no gene, ch2 POTEI off target\n",
    "# 18: of 7 results, 2 no gene, 1 is ACTB, 4 off target genes\n",
    "# 17: has ACTB, 3 no gene\n",
    "# 13: has ACTB, ACTG1 and OMIM genes on ch17, 2 no gene\n",
    "# 16: has ACTB, lots of overlapping genes\n",
    "\n",
    "# TSPAN6:\n",
    "# 257 has off target gene with correct TSPAN6\n",
    "# 249, 242, and 255 are good\n",
    "\n",
    "# None for EEF1A1\n",
    "\n",
    "# PTBP1:\n",
    "# 183, 186, 187, 188, 190\n",
    "# 185 is bad\n",
    "# 233, 210, 234, 219, 201, 216, 214, 236, 224 are bad\n",
    "\n",
    "# GCLC:\n",
    "# 156, 167, 169, 175, 177\n",
    "# 145, 176, 165 are bad\n",
    "\n",
    "# SCYL3:\n",
    "# 289 and 294 are bad\n",
    "# 268, 275, 276, 280, 284\n",
    "\n",
    "# CLCN6:\n",
    "# 106, 121, 109, 40 are bad\n",
    "# 46, 54, 102 have no gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "df1e6a84-4dff-41c1-ba57-746777dcc098",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> ID7\n",
      "ATCATGGGCAGTGAGCTCAGTGGTATCTGGA\n",
      "> ID0\n",
      "CTTTGGGGTTGCCCATGATGGCAGAGGCAG\n",
      "> ID12\n",
      "GGACAGGGCACTGGCCACTCCAGTCACCAT\n",
      "> ID6\n",
      "AGTTCACTCAGCTGGGCAAAGGTGCCCTTG\n",
      "> ID2\n",
      "CTTCTGCCAGGAAGCCTGCACCTCAGGGGT\n",
      "> ID8\n",
      "AGGACAGGTTGCCAAAGCTGTCAAAGAACCT\n"
     ]
    }
   ],
   "source": [
    "to_print = gene_probes_dict[\"NM_000184.3\"]\n",
    "for i, row in to_print.iterrows():\n",
    "    print(\"> ID\" + str(i))\n",
    "    print(row[\"sequence\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "b95b8b6c-468d-4c0c-87f3-94fbb8dbd09a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>sequence</th>\n",
       "      <th>Tm</th>\n",
       "      <th>on_target</th>\n",
       "      <th>off_target</th>\n",
       "      <th>repeat_seq</th>\n",
       "      <th>prob</th>\n",
       "      <th>max_kmer</th>\n",
       "      <th>probe_strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NM_000184</td>\n",
       "      <td>chr11</td>\n",
       "      <td>5253261</td>\n",
       "      <td>5253291</td>\n",
       "      <td>ATCATGGGCAGTGAGCTCAGTGGTATCTGGA</td>\n",
       "      <td>42.22</td>\n",
       "      <td>96.905</td>\n",
       "      <td>96.905</td>\n",
       "      <td>0</td>\n",
       "      <td>0.298</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_000184</td>\n",
       "      <td>HBG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_000184</td>\n",
       "      <td>chr11</td>\n",
       "      <td>5254426</td>\n",
       "      <td>5254455</td>\n",
       "      <td>CTTTGGGGTTGCCCATGATGGCAGAGGCAG</td>\n",
       "      <td>44.80</td>\n",
       "      <td>97.438</td>\n",
       "      <td>97.438</td>\n",
       "      <td>0</td>\n",
       "      <td>0.115</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_000184</td>\n",
       "      <td>HBG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NM_000184</td>\n",
       "      <td>chr11</td>\n",
       "      <td>5253292</td>\n",
       "      <td>5253321</td>\n",
       "      <td>GGACAGGGCACTGGCCACTCCAGTCACCAT</td>\n",
       "      <td>46.55</td>\n",
       "      <td>100.000</td>\n",
       "      <td>98.154</td>\n",
       "      <td>0</td>\n",
       "      <td>0.081</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_000184</td>\n",
       "      <td>HBG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NM_000184</td>\n",
       "      <td>chr11</td>\n",
       "      <td>5254332</td>\n",
       "      <td>5254361</td>\n",
       "      <td>AGTTCACTCAGCTGGGCAAAGGTGCCCTTG</td>\n",
       "      <td>44.24</td>\n",
       "      <td>99.200</td>\n",
       "      <td>99.200</td>\n",
       "      <td>0</td>\n",
       "      <td>0.079</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_000184</td>\n",
       "      <td>HBG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_000184</td>\n",
       "      <td>chr11</td>\n",
       "      <td>5253322</td>\n",
       "      <td>5253351</td>\n",
       "      <td>CTTCTGCCAGGAAGCCTGCACCTCAGGGGT</td>\n",
       "      <td>46.58</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.138</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_000184</td>\n",
       "      <td>HBG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NM_000184</td>\n",
       "      <td>chr11</td>\n",
       "      <td>5254456</td>\n",
       "      <td>5254486</td>\n",
       "      <td>AGGACAGGTTGCCAAAGCTGTCAAAGAACCT</td>\n",
       "      <td>42.08</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.137</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_000184</td>\n",
       "      <td>HBG2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       refseq  chrom    start     stop                         sequence  \\\n",
       "7   NM_000184  chr11  5253261  5253291  ATCATGGGCAGTGAGCTCAGTGGTATCTGGA   \n",
       "0   NM_000184  chr11  5254426  5254455   CTTTGGGGTTGCCCATGATGGCAGAGGCAG   \n",
       "12  NM_000184  chr11  5253292  5253321   GGACAGGGCACTGGCCACTCCAGTCACCAT   \n",
       "6   NM_000184  chr11  5254332  5254361   AGTTCACTCAGCTGGGCAAAGGTGCCCTTG   \n",
       "2   NM_000184  chr11  5253322  5253351   CTTCTGCCAGGAAGCCTGCACCTCAGGGGT   \n",
       "8   NM_000184  chr11  5254456  5254486  AGGACAGGTTGCCAAAGCTGTCAAAGAACCT   \n",
       "\n",
       "       Tm  on_target  off_target  repeat_seq   prob  max_kmer probe_strand  \\\n",
       "7   42.22     96.905      96.905           0  0.298         3            +   \n",
       "0   44.80     97.438      97.438           0  0.115         3            +   \n",
       "12  46.55    100.000      98.154           0  0.081         2            +   \n",
       "6   44.24     99.200      99.200           0  0.079         2            +   \n",
       "2   46.58    100.000     100.000           0  0.138         3            +   \n",
       "8   42.08    100.000     100.000           0  0.137         2            +   \n",
       "\n",
       "   transcript_id gene_id  \n",
       "7      NM_000184    HBG2  \n",
       "0      NM_000184    HBG2  \n",
       "12     NM_000184    HBG2  \n",
       "6      NM_000184    HBG2  \n",
       "2      NM_000184    HBG2  \n",
       "8      NM_000184    HBG2  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "to_print.head(n=70)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc93f5d6-e137-4116-869b-b5bd4aa981a0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1afadc4c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bc9b2a8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
