{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "d061f73f-f0f0-4582-9f4d-93aa4a847bc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Choosing genes for analysis ###\n",
    "\n",
    "### Citations of tools ###\n",
    "# gget: Luebbert, L., & Pachter, L. (2023). Efficient querying of genomic reference databases with gget. Bioinformatics. https://doi.org/10.1093/bioinformatics/btac836\n",
    "# pybiomart: https://github.com/jrderuiter/pybiomart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "b020c68d-137c-4b0a-ae2b-9a6db7450650",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import gget\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "4189d992-1ff8-41b2-9580-1c6aa9ee166a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data file downloaded from https://www.encodeproject.org/experiments/ENCSR792OIJ/\n",
    "# Accession ENCFF928NYA under processed data (tsv file) \n",
    "# Data added 2020-11-04\n",
    "df = pd.read_table(\"ENCFF928NYA.tsv\")\n",
    "columns = ['gene_id', 'TPM', 'FPKM']\n",
    "df = pd.DataFrame(df, columns=columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "2da92b03-41ba-4c84-8c64-89e972489124",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10904</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12954</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12956</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12958</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12960</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id  TPM  FPKM\n",
       "0   10904  0.0   0.0\n",
       "1   12954  0.0   0.0\n",
       "2   12956  0.0   0.0\n",
       "3   12958  0.0   0.0\n",
       "4   12960  0.0   0.0"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "b7f10d6e-ad37-4e81-978f-829bfd0cbb7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "expressed_df = df.query(\"TPM!=0.0\")\n",
    "\n",
    "high_thresh = 200\n",
    "high_thresh_df = expressed_df.query(\"TPM>@high_thresh\")\n",
    "\n",
    "low_thresh = 10\n",
    "low_thresh_df = expressed_df.query(\"TPM>5 & TPM<@low_thresh\")\n",
    "lowest_thresh_df = expressed_df.query(\"TPM<5\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "97db157c-ed5c-45c5-b250-ba93428c2aaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>13023</td>\n",
       "      <td>41.07</td>\n",
       "      <td>49.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>13024</td>\n",
       "      <td>16.31</td>\n",
       "      <td>19.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>30958</td>\n",
       "      <td>43.49</td>\n",
       "      <td>52.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471</th>\n",
       "      <td>30964</td>\n",
       "      <td>52.81</td>\n",
       "      <td>63.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_id    TPM   FPKM\n",
       "46                13023  41.07  49.61\n",
       "47                13024  16.31  19.70\n",
       "465               30958  43.49  52.53\n",
       "471               30964  52.81  63.79\n",
       "649  ENSG00000000003.14   0.05   0.06"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expressed_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "52f533ef-ba01-4c2d-a155-beaa77db2d5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rows were selected by looking at the entire dataframe and were chosen for varying expression levels\n",
    "high_rows = [941, 1965, 10644, 17808, 24929, 48356, 53485, 54599, 55026, 55115, 57853]\n",
    "low_rows = [657, 926]\n",
    "lowest_rows = [649, 652]\n",
    "\n",
    "chosen_rows = high_rows + low_rows + lowest_rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "333b81a4-0dba-4157-ab5f-55e7d0584c4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>ENSG00000011304.19</td>\n",
       "      <td>342.64</td>\n",
       "      <td>413.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1965</th>\n",
       "      <td>ENSG00000075624.14</td>\n",
       "      <td>1567.70</td>\n",
       "      <td>1893.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10644</th>\n",
       "      <td>ENSG00000156508.17</td>\n",
       "      <td>5000.06</td>\n",
       "      <td>6039.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17808</th>\n",
       "      <td>ENSG00000196565.14</td>\n",
       "      <td>4169.72</td>\n",
       "      <td>5036.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24929</th>\n",
       "      <td>ENSG00000222328.1</td>\n",
       "      <td>7046.27</td>\n",
       "      <td>8511.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48356</th>\n",
       "      <td>ENSG00000263934.4</td>\n",
       "      <td>13928.59</td>\n",
       "      <td>16824.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53485</th>\n",
       "      <td>ENSG00000274012.1</td>\n",
       "      <td>94300.79</td>\n",
       "      <td>113903.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54599</th>\n",
       "      <td>ENSG00000276168.1</td>\n",
       "      <td>149831.97</td>\n",
       "      <td>180978.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55026</th>\n",
       "      <td>ENSG00000277027.1</td>\n",
       "      <td>23024.05</td>\n",
       "      <td>27810.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55115</th>\n",
       "      <td>ENSG00000277209.1</td>\n",
       "      <td>43982.92</td>\n",
       "      <td>53125.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57853</th>\n",
       "      <td>ENSG00000283293.1</td>\n",
       "      <td>66929.44</td>\n",
       "      <td>80842.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657</th>\n",
       "      <td>ENSG00000001084.12</td>\n",
       "      <td>6.26</td>\n",
       "      <td>7.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926</th>\n",
       "      <td>ENSG00000011021.22</td>\n",
       "      <td>9.82</td>\n",
       "      <td>11.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>ENSG00000000457.13</td>\n",
       "      <td>3.32</td>\n",
       "      <td>4.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_id        TPM       FPKM\n",
       "941    ENSG00000011304.19     342.64     413.86\n",
       "1965   ENSG00000075624.14    1567.70    1893.59\n",
       "10644  ENSG00000156508.17    5000.06    6039.46\n",
       "17808  ENSG00000196565.14    4169.72    5036.51\n",
       "24929   ENSG00000222328.1    7046.27    8511.03\n",
       "48356   ENSG00000263934.4   13928.59   16824.04\n",
       "53485   ENSG00000274012.1   94300.79  113903.82\n",
       "54599   ENSG00000276168.1  149831.97  180978.68\n",
       "55026   ENSG00000277027.1   23024.05   27810.24\n",
       "55115   ENSG00000277209.1   43982.92   53125.98\n",
       "57853   ENSG00000283293.1   66929.44   80842.58\n",
       "657    ENSG00000001084.12       6.26       7.56\n",
       "926    ENSG00000011021.22       9.82      11.86\n",
       "649    ENSG00000000003.14       0.05       0.06\n",
       "652    ENSG00000000457.13       3.32       4.01"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chosen_rows_df = pd.DataFrame(df.loc[chosen_rows])\n",
    "\n",
    "chosen_rows_gene_ids = []\n",
    "for index, row in chosen_rows_df.iterrows():\n",
    "    chosen_rows_gene_ids.append(row['gene_id'])\n",
    "    \n",
    "chosen_rows_df.head(n=len(chosen_rows_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "e365a725-64c3-4883-afc6-21d00abd6211",
   "metadata": {},
   "outputs": [],
   "source": [
    "### METHOD USING GGET ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "db428d50-2607-444d-a4e9-7d40453b62e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Tue Jun 27 20:57:56 2023 INFO We noticed that you passed a version number with your Ensembl ID.\n",
      "Please note that gget info will always return information linked to the latest Ensembl ID version (see 'ensembl_id').\n"
     ]
    }
   ],
   "source": [
    "res = gget.info(chosen_rows_gene_ids, json=True, uniprot=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "13274da2-9161-48ca-ab97-4c46fd657101",
   "metadata": {},
   "outputs": [],
   "source": [
    "ordered_gene_names = [None] * len(chosen_rows_gene_ids)\n",
    "\n",
    "for key in res.keys():\n",
    "    for i in range(len(chosen_rows_gene_ids)):\n",
    "        if chosen_rows_gene_ids[i].startswith(key):\n",
    "            ordered_gene_names[i] = res[key]['ensembl_gene_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "7ac14023-a8ad-4058-8bc2-917d011e83c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene name (ensembl)</th>\n",
       "      <th>gene name (ensembl)</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>ENSG00000011304.19</td>\n",
       "      <td>PTBP1</td>\n",
       "      <td>PTBP1</td>\n",
       "      <td>342.64</td>\n",
       "      <td>413.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1965</th>\n",
       "      <td>ENSG00000075624.14</td>\n",
       "      <td>ACTB</td>\n",
       "      <td>ACTB</td>\n",
       "      <td>1567.70</td>\n",
       "      <td>1893.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10644</th>\n",
       "      <td>ENSG00000156508.17</td>\n",
       "      <td>EEF1A1</td>\n",
       "      <td>EEF1A1</td>\n",
       "      <td>5000.06</td>\n",
       "      <td>6039.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17808</th>\n",
       "      <td>ENSG00000196565.14</td>\n",
       "      <td>HBG2</td>\n",
       "      <td>HBG2</td>\n",
       "      <td>4169.72</td>\n",
       "      <td>5036.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24929</th>\n",
       "      <td>ENSG00000222328.1</td>\n",
       "      <td>RNU2-2P</td>\n",
       "      <td>RNU2-2P</td>\n",
       "      <td>7046.27</td>\n",
       "      <td>8511.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48356</th>\n",
       "      <td>ENSG00000263934.4</td>\n",
       "      <td>SNORD3A</td>\n",
       "      <td>SNORD3A</td>\n",
       "      <td>13928.59</td>\n",
       "      <td>16824.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53485</th>\n",
       "      <td>ENSG00000274012.1</td>\n",
       "      <td>RN7SL2</td>\n",
       "      <td>RN7SL2</td>\n",
       "      <td>94300.79</td>\n",
       "      <td>113903.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54599</th>\n",
       "      <td>ENSG00000276168.1</td>\n",
       "      <td>RN7SL1</td>\n",
       "      <td>RN7SL1</td>\n",
       "      <td>149831.97</td>\n",
       "      <td>180978.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55026</th>\n",
       "      <td>ENSG00000277027.1</td>\n",
       "      <td>RMRP</td>\n",
       "      <td>RMRP</td>\n",
       "      <td>23024.05</td>\n",
       "      <td>27810.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55115</th>\n",
       "      <td>ENSG00000277209.1</td>\n",
       "      <td>RPPH1</td>\n",
       "      <td>RPPH1</td>\n",
       "      <td>43982.92</td>\n",
       "      <td>53125.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57853</th>\n",
       "      <td>ENSG00000283293.1</td>\n",
       "      <td>RN7SK</td>\n",
       "      <td>RN7SK</td>\n",
       "      <td>66929.44</td>\n",
       "      <td>80842.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657</th>\n",
       "      <td>ENSG00000001084.12</td>\n",
       "      <td>GCLC</td>\n",
       "      <td>GCLC</td>\n",
       "      <td>6.26</td>\n",
       "      <td>7.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926</th>\n",
       "      <td>ENSG00000011021.22</td>\n",
       "      <td>CLCN6</td>\n",
       "      <td>CLCN6</td>\n",
       "      <td>9.82</td>\n",
       "      <td>11.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>ENSG00000000457.13</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>3.32</td>\n",
       "      <td>4.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_id gene name (ensembl) gene name (ensembl)        TPM  \\\n",
       "941    ENSG00000011304.19               PTBP1               PTBP1     342.64   \n",
       "1965   ENSG00000075624.14                ACTB                ACTB    1567.70   \n",
       "10644  ENSG00000156508.17              EEF1A1              EEF1A1    5000.06   \n",
       "17808  ENSG00000196565.14                HBG2                HBG2    4169.72   \n",
       "24929   ENSG00000222328.1             RNU2-2P             RNU2-2P    7046.27   \n",
       "48356   ENSG00000263934.4             SNORD3A             SNORD3A   13928.59   \n",
       "53485   ENSG00000274012.1              RN7SL2              RN7SL2   94300.79   \n",
       "54599   ENSG00000276168.1              RN7SL1              RN7SL1  149831.97   \n",
       "55026   ENSG00000277027.1                RMRP                RMRP   23024.05   \n",
       "55115   ENSG00000277209.1               RPPH1               RPPH1   43982.92   \n",
       "57853   ENSG00000283293.1               RN7SK               RN7SK   66929.44   \n",
       "657    ENSG00000001084.12                GCLC                GCLC       6.26   \n",
       "926    ENSG00000011021.22               CLCN6               CLCN6       9.82   \n",
       "649    ENSG00000000003.14              TSPAN6              TSPAN6       0.05   \n",
       "652    ENSG00000000457.13               SCYL3               SCYL3       3.32   \n",
       "\n",
       "            FPKM  \n",
       "941       413.86  \n",
       "1965     1893.59  \n",
       "10644    6039.46  \n",
       "17808    5036.51  \n",
       "24929    8511.03  \n",
       "48356   16824.04  \n",
       "53485  113903.82  \n",
       "54599  180978.68  \n",
       "55026   27810.24  \n",
       "55115   53125.98  \n",
       "57853   80842.58  \n",
       "657         7.56  \n",
       "926        11.86  \n",
       "649         0.06  \n",
       "652         4.01  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chosen_rows_df.insert(1, \"gene name (ensembl)\", ordered_gene_names, True)\n",
    "chosen_rows_df.head(n=len(chosen_rows_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "6f55cf83-80cd-45ac-bc78-cc4e92fb9897",
   "metadata": {},
   "outputs": [],
   "source": [
    "# However, we need the RefSeq ID beginning in NM_ to use PaintSHOP, and gget doesn't let us do that. \n",
    "# So, the following method uses pybiomart which uses the ensembl API and MANE which correlates RefSeq IDs to Ensembl IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "d50ead79-9d54-4f09-83a5-efd435f319e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pybiomart import Dataset\n",
    "\n",
    "dataset = Dataset(name='hsapiens_gene_ensembl',\n",
    "                  host='http://www.ensembl.org')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "75c00436-8cd8-4b94-a898-b1a6dcade9a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "refseq_results = []\n",
    "\n",
    "# Gets the gene name and the RefSeq ID associated with the canonical (most common) transcript\n",
    "for i in chosen_rows_gene_ids:\n",
    "    gene_id = i.split(\".\")[0] # remove the version number\n",
    "    result = dataset.query(attributes=['ensembl_gene_id', 'external_gene_name', 'external_transcript_name', 'transcript_mane_select', 'transcript_is_canonical', 'ensembl_transcript_id'],\n",
    "                  filters={'link_ensembl_gene_id': [gene_id], 'transcript_is_canonical' : True})\n",
    "    refseq_results.append(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "603f2fb5-fe6f-4260-adbc-f0fdf3060772",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Gene stable ID Gene name Transcript name RefSeq match transcript (MANE Select)  Ensembl Canonical Transcript stable ID\n",
      "0   ENSG00000011304     PTBP1       PTBP1-203                           NM_002819.5               True      ENST00000356948\n",
      "1   ENSG00000075624      ACTB        ACTB-217                           NM_001101.5               True      ENST00000646664\n",
      "2   ENSG00000156508    EEF1A1      EEF1A1-201                           NM_001402.6               True      ENST00000309268\n",
      "3   ENSG00000196565      HBG2        HBG2-201                           NM_000184.3               True      ENST00000336906\n",
      "4   ENSG00000222328   RNU2-2P     RNU2-2P-201                                   NaN               True      ENST00000410396\n",
      "5   ENSG00000263934   SNORD3A     SNORD3A-201                                   NaN               True      ENST00000584923\n",
      "6   ENSG00000274012    RN7SL2      RN7SL2-201                                   NaN               True      ENST00000490232\n",
      "7   ENSG00000276168    RN7SL1      RN7SL1-201                                   NaN               True      ENST00000618786\n",
      "8   ENSG00000277027      RMRP        RMRP-201                                   NaN               True      ENST00000363046\n",
      "9   ENSG00000277209     RPPH1       RPPH1-201                                   NaN               True      ENST00000516869\n",
      "10  ENSG00000283293     RN7SK       RN7SK-201                                   NaN               True      ENST00000636484\n",
      "11  ENSG00000001084      GCLC        GCLC-214                           NM_001498.4               True      ENST00000650454\n",
      "12  ENSG00000011021     CLCN6       CLCN6-202                           NM_001286.5               True      ENST00000346436\n",
      "13  ENSG00000000003    TSPAN6      TSPAN6-201                           NM_003270.4               True      ENST00000373020\n",
      "14  ENSG00000000457     SCYL3       SCYL3-202                           NM_020423.7               True      ENST00000367771\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Transcript name</th>\n",
       "      <th>RefSeq match transcript (MANE Select)</th>\n",
       "      <th>Ensembl Canonical</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000011304</td>\n",
       "      <td>PTBP1</td>\n",
       "      <td>PTBP1-203</td>\n",
       "      <td>NM_002819.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000356948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000075624</td>\n",
       "      <td>ACTB</td>\n",
       "      <td>ACTB-217</td>\n",
       "      <td>NM_001101.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000646664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000156508</td>\n",
       "      <td>EEF1A1</td>\n",
       "      <td>EEF1A1-201</td>\n",
       "      <td>NM_001402.6</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000309268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000196565</td>\n",
       "      <td>HBG2</td>\n",
       "      <td>HBG2-201</td>\n",
       "      <td>NM_000184.3</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000336906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000222328</td>\n",
       "      <td>RNU2-2P</td>\n",
       "      <td>RNU2-2P-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000410396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000263934</td>\n",
       "      <td>SNORD3A</td>\n",
       "      <td>SNORD3A-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000584923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000274012</td>\n",
       "      <td>RN7SL2</td>\n",
       "      <td>RN7SL2-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000490232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000276168</td>\n",
       "      <td>RN7SL1</td>\n",
       "      <td>RN7SL1-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000618786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000277027</td>\n",
       "      <td>RMRP</td>\n",
       "      <td>RMRP-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000363046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000277209</td>\n",
       "      <td>RPPH1</td>\n",
       "      <td>RPPH1-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000516869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000283293</td>\n",
       "      <td>RN7SK</td>\n",
       "      <td>RN7SK-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000636484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000001084</td>\n",
       "      <td>GCLC</td>\n",
       "      <td>GCLC-214</td>\n",
       "      <td>NM_001498.4</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000650454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000011021</td>\n",
       "      <td>CLCN6</td>\n",
       "      <td>CLCN6-202</td>\n",
       "      <td>NM_001286.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000346436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>TSPAN6-201</td>\n",
       "      <td>NM_003270.4</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000373020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>SCYL3-202</td>\n",
       "      <td>NM_020423.7</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000367771</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene stable ID Gene name Transcript name  \\\n",
       "0   ENSG00000011304     PTBP1       PTBP1-203   \n",
       "1   ENSG00000075624      ACTB        ACTB-217   \n",
       "2   ENSG00000156508    EEF1A1      EEF1A1-201   \n",
       "3   ENSG00000196565      HBG2        HBG2-201   \n",
       "4   ENSG00000222328   RNU2-2P     RNU2-2P-201   \n",
       "5   ENSG00000263934   SNORD3A     SNORD3A-201   \n",
       "6   ENSG00000274012    RN7SL2      RN7SL2-201   \n",
       "7   ENSG00000276168    RN7SL1      RN7SL1-201   \n",
       "8   ENSG00000277027      RMRP        RMRP-201   \n",
       "9   ENSG00000277209     RPPH1       RPPH1-201   \n",
       "10  ENSG00000283293     RN7SK       RN7SK-201   \n",
       "11  ENSG00000001084      GCLC        GCLC-214   \n",
       "12  ENSG00000011021     CLCN6       CLCN6-202   \n",
       "13  ENSG00000000003    TSPAN6      TSPAN6-201   \n",
       "14  ENSG00000000457     SCYL3       SCYL3-202   \n",
       "\n",
       "   RefSeq match transcript (MANE Select)  Ensembl Canonical  \\\n",
       "0                            NM_002819.5               True   \n",
       "1                            NM_001101.5               True   \n",
       "2                            NM_001402.6               True   \n",
       "3                            NM_000184.3               True   \n",
       "4                                    NaN               True   \n",
       "5                                    NaN               True   \n",
       "6                                    NaN               True   \n",
       "7                                    NaN               True   \n",
       "8                                    NaN               True   \n",
       "9                                    NaN               True   \n",
       "10                                   NaN               True   \n",
       "11                           NM_001498.4               True   \n",
       "12                           NM_001286.5               True   \n",
       "13                           NM_003270.4               True   \n",
       "14                           NM_020423.7               True   \n",
       "\n",
       "   Transcript stable ID  \n",
       "0       ENST00000356948  \n",
       "1       ENST00000646664  \n",
       "2       ENST00000309268  \n",
       "3       ENST00000336906  \n",
       "4       ENST00000410396  \n",
       "5       ENST00000584923  \n",
       "6       ENST00000490232  \n",
       "7       ENST00000618786  \n",
       "8       ENST00000363046  \n",
       "9       ENST00000516869  \n",
       "10      ENST00000636484  \n",
       "11      ENST00000650454  \n",
       "12      ENST00000346436  \n",
       "13      ENST00000373020  \n",
       "14      ENST00000367771  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_df = pd.concat(refseq_results, ignore_index = True)\n",
    "refseq_df['Ensembl Canonical'] = refseq_df['Ensembl Canonical'].astype(bool)  \n",
    "print(refseq_df.to_string())\n",
    "refseq_df.head(n=len(refseq_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "96642316-9280-4da2-810c-91ea0d1717a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note that several of the genes have no matching RefSeq ID. This is because they are not mRNA, but other kinds of RNA (i.e. snRNA or rRNA). Ensembl doesn't already have the RefSeq IDs available for this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "810ed68f-256b-4872-b5b6-6881b50ffd9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export to file to use in PaintSHOP\n",
    "f = open(\"refseqs_sample.txt\", \"w\")\n",
    "for index, row in refseq_df.iterrows():\n",
    "    if str(row['RefSeq match transcript (MANE Select)']) != 'nan':\n",
    "        f.write(str(row['RefSeq match transcript (MANE Select)']) + '\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "08adaab0-cb19-4bc7-96c8-2d5191004956",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Then we use PaintSHOP to make probes: https://doi.org/10.1038/s41592-021-01187-3"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
