{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8eb96018-d837-4ec0-b0ab-ee04079195b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)\n",
    "import gget\n",
    "import json\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "55c60eb4-f6e3-4dd0-86cf-e5b16a186781",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data file downloaded from https://www.encodeproject.org/experiments/ENCSR792OIJ/\n",
    "# Accession ENCFF928NYA under processed data (tsv file) \n",
    "# Data added 2020-11-04\n",
    "df = pd.read_table(\"ENCFF928NYA.tsv\")\n",
    "columns = ['gene_id', 'TPM', 'FPKM']\n",
    "df = pd.DataFrame(df, columns=columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3c8776a0-b4d6-41c8-a46c-0d5c9daa59e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "neg_ctrl = df.query(\"gene_id=='ENSG00000057149.15' | gene_id=='ENSG00000206073.10'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f8e0d76d-3f6d-4746-98fc-cc1b91bda955",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                  gene_id  TPM  FPKM\n",
      "1418   ENSG00000057149.15  0.0   0.0\n",
      "21010  ENSG00000206073.10  0.0   0.0\n"
     ]
    }
   ],
   "source": [
    "print(neg_ctrl.to_string())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b0846b55-b2fd-4a15-bab1-987313a85746",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pybiomart import Dataset\n",
    "\n",
    "dataset = Dataset(name='hsapiens_gene_ensembl',\n",
    "                  host='http://www.ensembl.org')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "86084c5a-e66b-4f1f-a43e-9a59ac9bf3cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "refseq_results = []\n",
    "\n",
    "# Gets the gene name and the RefSeq ID associated with the canonical (most common) transcript\n",
    "for i, row in neg_ctrl.iterrows():\n",
    "    gene_id = row[\"gene_id\"].split(\".\")[0] # remove the version number\n",
    "    # print(gene_id)\n",
    "    result = dataset.query(attributes=['ensembl_gene_id', 'external_gene_name', 'external_transcript_name', 'transcript_mane_select', 'transcript_is_canonical', 'ensembl_transcript_id'],\n",
    "                  filters={'link_ensembl_gene_id': [gene_id], 'transcript_is_canonical' : True})\n",
    "    result[\"gene_id\"] = row[\"gene_id\"]\n",
    "    refseq_results.append(result)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "71ba86db-93b5-4ae4-a515-608643793463",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Gene stable ID Gene name Transcript name RefSeq match transcript (MANE Select)  Ensembl Canonical Transcript stable ID             gene_id\n",
      "0  ENSG00000057149  SERPINB3    SERPINB3-201                           NM_006919.3               True      ENST00000283752  ENSG00000057149.15\n",
      "1  ENSG00000206073  SERPINB4    SERPINB4-201                           NM_002974.4               True      ENST00000341074  ENSG00000206073.10\n"
     ]
    }
   ],
   "source": [
    "refseq_df = pd.concat(refseq_results, ignore_index = True)\n",
    "refseq_df['Ensembl Canonical'] = refseq_df['Ensembl Canonical'].astype(bool)\n",
    "print(refseq_df.to_string())\n",
    "# refseq_df.head(n=len(refseq_df.index))\n",
    "\n",
    "# for i, row in "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "98688b64-e83e-4bec-a7c8-d44c4b2e3890",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Transcript name</th>\n",
       "      <th>RefSeq match transcript (MANE Select)</th>\n",
       "      <th>Ensembl Canonical</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000057149.15</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENSG00000057149</td>\n",
       "      <td>SERPINB3</td>\n",
       "      <td>SERPINB3-201</td>\n",
       "      <td>NM_006919.3</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000283752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000206073.10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENSG00000206073</td>\n",
       "      <td>SERPINB4</td>\n",
       "      <td>SERPINB4-201</td>\n",
       "      <td>NM_002974.4</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000341074</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id  TPM  FPKM   Gene stable ID Gene name Transcript name  \\\n",
       "0  ENSG00000057149.15  0.0   0.0  ENSG00000057149  SERPINB3    SERPINB3-201   \n",
       "1  ENSG00000206073.10  0.0   0.0  ENSG00000206073  SERPINB4    SERPINB4-201   \n",
       "\n",
       "  RefSeq match transcript (MANE Select)  Ensembl Canonical  \\\n",
       "0                           NM_006919.3               True   \n",
       "1                           NM_002974.4               True   \n",
       "\n",
       "  Transcript stable ID  \n",
       "0      ENST00000283752  \n",
       "1      ENST00000341074  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedRes = pd.merge(neg_ctrl, refseq_df, on ='gene_id')\n",
    "mergedRes.head(len(mergedRes.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "a0225b47-0816-4001-a2ef-07af341b6f88",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_table('230630-SERPINB3-probes.txt')\n",
    "#data = data.sort_values(['off_target', 'prob'], ascending=[True, False])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "b4c794dd-4151-4a9b-8b60-15af28a8a46a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>sequence</th>\n",
       "      <th>Tm</th>\n",
       "      <th>on_target</th>\n",
       "      <th>off_target</th>\n",
       "      <th>repeat_seq</th>\n",
       "      <th>prob</th>\n",
       "      <th>max_kmer</th>\n",
       "      <th>probe_strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63655754</td>\n",
       "      <td>63655790</td>\n",
       "      <td>GTTGAAGTAGGTGATGATCCGAATCCTACTACAGCGG</td>\n",
       "      <td>42.02</td>\n",
       "      <td>100.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.264</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63656915</td>\n",
       "      <td>63656951</td>\n",
       "      <td>CTGTACATCCTCCAGCGAGGCAAAATGAAAAGATGTG</td>\n",
       "      <td>42.22</td>\n",
       "      <td>99.986</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.159</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63655853</td>\n",
       "      <td>63655882</td>\n",
       "      <td>GATAGCACGAGACCGCGGCTCCCGGTCATG</td>\n",
       "      <td>46.69</td>\n",
       "      <td>98.078</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.157</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63655791</td>\n",
       "      <td>63655820</td>\n",
       "      <td>TGGCAGCTGCAGCTTCTGCTCCCTCCTCTG</td>\n",
       "      <td>46.62</td>\n",
       "      <td>99.251</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.153</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63655603</td>\n",
       "      <td>63655639</td>\n",
       "      <td>ACCAGAACATCTGCAGGTGAACATTTTCCAAATGGAG</td>\n",
       "      <td>42.14</td>\n",
       "      <td>100.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.096</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63655960</td>\n",
       "      <td>63655996</td>\n",
       "      <td>TTCCACTTTGAACCGAGGTAAGTGTAAATCGACACGT</td>\n",
       "      <td>42.28</td>\n",
       "      <td>96.532</td>\n",
       "      <td>61.311</td>\n",
       "      <td>0</td>\n",
       "      <td>0.203</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63658560</td>\n",
       "      <td>63658596</td>\n",
       "      <td>TCTGGAGCATTTGCAAAATCAACAGATTCCACACTGG</td>\n",
       "      <td>42.41</td>\n",
       "      <td>100.000</td>\n",
       "      <td>79.944</td>\n",
       "      <td>0</td>\n",
       "      <td>0.089</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63661156</td>\n",
       "      <td>63661192</td>\n",
       "      <td>TTCTGAACTGTTGGAACAGGTCGAACATGAACTTGGT</td>\n",
       "      <td>42.66</td>\n",
       "      <td>100.000</td>\n",
       "      <td>81.777</td>\n",
       "      <td>0</td>\n",
       "      <td>0.090</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63655883</td>\n",
       "      <td>63655916</td>\n",
       "      <td>CCTGAGAGGTCTGCATCCCCATTGAAGATATCCA</td>\n",
       "      <td>42.05</td>\n",
       "      <td>94.174</td>\n",
       "      <td>91.286</td>\n",
       "      <td>0</td>\n",
       "      <td>0.142</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63655821</td>\n",
       "      <td>63655852</td>\n",
       "      <td>TAACCTCCACAAAGGCCTTGTGTAGGACTCCA</td>\n",
       "      <td>42.17</td>\n",
       "      <td>100.000</td>\n",
       "      <td>93.235</td>\n",
       "      <td>0</td>\n",
       "      <td>0.090</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63660801</td>\n",
       "      <td>63660837</td>\n",
       "      <td>TGATATGTTGCAGCTTTTCCTGTGGTGTTCTCTGTGA</td>\n",
       "      <td>42.61</td>\n",
       "      <td>100.000</td>\n",
       "      <td>94.371</td>\n",
       "      <td>0</td>\n",
       "      <td>0.338</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63661089</td>\n",
       "      <td>63661122</td>\n",
       "      <td>AAGAGGACCATCCCTAATGCTGATGTGATGCTGA</td>\n",
       "      <td>42.17</td>\n",
       "      <td>95.953</td>\n",
       "      <td>95.953</td>\n",
       "      <td>0</td>\n",
       "      <td>0.214</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63655640</td>\n",
       "      <td>63655676</td>\n",
       "      <td>TGACAGACTAATTGCATCTACGGGGATGAGAATCTGC</td>\n",
       "      <td>42.12</td>\n",
       "      <td>100.000</td>\n",
       "      <td>97.434</td>\n",
       "      <td>0</td>\n",
       "      <td>0.139</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63657316</td>\n",
       "      <td>63657352</td>\n",
       "      <td>TTCTTCTCCCACTGCCCTTTGAAATAGATTGCGTTCA</td>\n",
       "      <td>42.69</td>\n",
       "      <td>100.000</td>\n",
       "      <td>99.791</td>\n",
       "      <td>0</td>\n",
       "      <td>0.268</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63661193</td>\n",
       "      <td>63661229</td>\n",
       "      <td>GTTGGCTTCACTGAGTGAATTCATGGTGAACTCGATG</td>\n",
       "      <td>42.13</td>\n",
       "      <td>99.793</td>\n",
       "      <td>99.793</td>\n",
       "      <td>0</td>\n",
       "      <td>0.144</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63655917</td>\n",
       "      <td>63655948</td>\n",
       "      <td>CCATTCCCATGGTTCTCAACGTGTCCTTGAGG</td>\n",
       "      <td>42.19</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.249</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63656878</td>\n",
       "      <td>63656914</td>\n",
       "      <td>GATCTTTGCCTTTGTATGGTATTTCCAGGACCTTGGC</td>\n",
       "      <td>42.08</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.249</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63659418</td>\n",
       "      <td>63659450</td>\n",
       "      <td>TTTTCTCCGAAGAGCTTGTTGGCGATCTTCAGC</td>\n",
       "      <td>42.29</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.190</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NM_006919</td>\n",
       "      <td>chr18</td>\n",
       "      <td>63655564</td>\n",
       "      <td>63655600</td>\n",
       "      <td>TATGAGCCAAGAGAATCTGTTGTTGCCAGCAATCAGT</td>\n",
       "      <td>42.72</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.091</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_006919</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       refseq  chrom     start      stop  \\\n",
       "4   NM_006919  chr18  63655754  63655790   \n",
       "16  NM_006919  chr18  63656915  63656951   \n",
       "3   NM_006919  chr18  63655853  63655882   \n",
       "5   NM_006919  chr18  63655791  63655820   \n",
       "0   NM_006919  chr18  63655603  63655639   \n",
       "14  NM_006919  chr18  63655960  63655996   \n",
       "18  NM_006919  chr18  63658560  63658596   \n",
       "10  NM_006919  chr18  63661156  63661192   \n",
       "12  NM_006919  chr18  63655883  63655916   \n",
       "2   NM_006919  chr18  63655821  63655852   \n",
       "8   NM_006919  chr18  63660801  63660837   \n",
       "9   NM_006919  chr18  63661089  63661122   \n",
       "1   NM_006919  chr18  63655640  63655676   \n",
       "17  NM_006919  chr18  63657316  63657352   \n",
       "11  NM_006919  chr18  63661193  63661229   \n",
       "13  NM_006919  chr18  63655917  63655948   \n",
       "15  NM_006919  chr18  63656878  63656914   \n",
       "7   NM_006919  chr18  63659418  63659450   \n",
       "6   NM_006919  chr18  63655564  63655600   \n",
       "\n",
       "                                 sequence     Tm  on_target  off_target  \\\n",
       "4   GTTGAAGTAGGTGATGATCCGAATCCTACTACAGCGG  42.02    100.000       0.000   \n",
       "16  CTGTACATCCTCCAGCGAGGCAAAATGAAAAGATGTG  42.22     99.986       0.000   \n",
       "3          GATAGCACGAGACCGCGGCTCCCGGTCATG  46.69     98.078       0.000   \n",
       "5          TGGCAGCTGCAGCTTCTGCTCCCTCCTCTG  46.62     99.251       0.000   \n",
       "0   ACCAGAACATCTGCAGGTGAACATTTTCCAAATGGAG  42.14    100.000       0.000   \n",
       "14  TTCCACTTTGAACCGAGGTAAGTGTAAATCGACACGT  42.28     96.532      61.311   \n",
       "18  TCTGGAGCATTTGCAAAATCAACAGATTCCACACTGG  42.41    100.000      79.944   \n",
       "10  TTCTGAACTGTTGGAACAGGTCGAACATGAACTTGGT  42.66    100.000      81.777   \n",
       "12     CCTGAGAGGTCTGCATCCCCATTGAAGATATCCA  42.05     94.174      91.286   \n",
       "2        TAACCTCCACAAAGGCCTTGTGTAGGACTCCA  42.17    100.000      93.235   \n",
       "8   TGATATGTTGCAGCTTTTCCTGTGGTGTTCTCTGTGA  42.61    100.000      94.371   \n",
       "9      AAGAGGACCATCCCTAATGCTGATGTGATGCTGA  42.17     95.953      95.953   \n",
       "1   TGACAGACTAATTGCATCTACGGGGATGAGAATCTGC  42.12    100.000      97.434   \n",
       "17  TTCTTCTCCCACTGCCCTTTGAAATAGATTGCGTTCA  42.69    100.000      99.791   \n",
       "11  GTTGGCTTCACTGAGTGAATTCATGGTGAACTCGATG  42.13     99.793      99.793   \n",
       "13       CCATTCCCATGGTTCTCAACGTGTCCTTGAGG  42.19    100.000     100.000   \n",
       "15  GATCTTTGCCTTTGTATGGTATTTCCAGGACCTTGGC  42.08    100.000     100.000   \n",
       "7       TTTTCTCCGAAGAGCTTGTTGGCGATCTTCAGC  42.29    100.000     100.000   \n",
       "6   TATGAGCCAAGAGAATCTGTTGTTGCCAGCAATCAGT  42.72    100.000     100.000   \n",
       "\n",
       "    repeat_seq   prob  max_kmer probe_strand transcript_id   gene_id  \n",
       "4            0  0.264         0            +     NM_006919  SERPINB3  \n",
       "16           0  0.159         2            +     NM_006919  SERPINB3  \n",
       "3            0  0.157         0            +     NM_006919  SERPINB3  \n",
       "5            0  0.153         3            +     NM_006919  SERPINB3  \n",
       "0            0  0.096         3            +     NM_006919  SERPINB3  \n",
       "14           0  0.203         2            +     NM_006919  SERPINB3  \n",
       "18           0  0.089         2            +     NM_006919  SERPINB3  \n",
       "10           0  0.090         2            +     NM_006919  SERPINB3  \n",
       "12           0  0.142         2            +     NM_006919  SERPINB3  \n",
       "2            0  0.090         2            +     NM_006919  SERPINB3  \n",
       "8            0  0.338         3            +     NM_006919  SERPINB3  \n",
       "9            0  0.214         3            +     NM_006919  SERPINB3  \n",
       "1            0  0.139         2            +     NM_006919  SERPINB3  \n",
       "17           0  0.268         3            +     NM_006919  SERPINB3  \n",
       "11           0  0.144         5            +     NM_006919  SERPINB3  \n",
       "13           0  0.249         2            +     NM_006919  SERPINB3  \n",
       "15           0  0.249         3            +     NM_006919  SERPINB3  \n",
       "7            0  0.190         2            +     NM_006919  SERPINB3  \n",
       "6            0  0.091         2            +     NM_006919  SERPINB3  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mask = (data['sequence'].str.len() <= 37)\n",
    "data = data.loc[mask]\n",
    "data = data.sort_values(['off_target', 'prob'], ascending=[True, False])\n",
    "data.head(n=300)\n",
    "\n",
    "#data.query(\"'gene_id'=='EEF1A1'\").head(n=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "aee87af7-d9b9-42a8-9aa2-2dde935fa667",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'>\n",
      "NM_006919 19\n"
     ]
    }
   ],
   "source": [
    "gene_probes_dict = {}\n",
    "valid_seqs = ['NM_006919']\n",
    "\n",
    "for i in valid_seqs:\n",
    "    gene = data.query(\"refseq=='\" + i.split(\".\")[0] + \"'\")\n",
    "    gene_probes_dict[i] = gene\n",
    "    print(type(\"refseq=='\" + i.split(\".\")[0] + \"'\"))\n",
    "    print(i, len(gene.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "5a05a449-1cc4-4c88-bb18-74d18fe3506f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> ID0\n",
      "ACCAGAACATCTGCAGGTGAACATTTTCCAAATGGAG\n",
      "> ID3\n",
      "GATAGCACGAGACCGCGGCTCCCGGTCATG\n",
      "> ID4\n",
      "GTTGAAGTAGGTGATGATCCGAATCCTACTACAGCGG\n",
      "> ID5\n",
      "TGGCAGCTGCAGCTTCTGCTCCCTCCTCTG\n",
      "> ID16\n",
      "CTGTACATCCTCCAGCGAGGCAAAATGAAAAGATGTG\n",
      "> ID14\n",
      "TTCCACTTTGAACCGAGGTAAGTGTAAATCGACACGT\n",
      "> ID18\n",
      "TCTGGAGCATTTGCAAAATCAACAGATTCCACACTGG\n",
      "> ID10\n",
      "TTCTGAACTGTTGGAACAGGTCGAACATGAACTTGGT\n",
      "> ID12\n",
      "CCTGAGAGGTCTGCATCCCCATTGAAGATATCCA\n",
      "> ID2\n",
      "TAACCTCCACAAAGGCCTTGTGTAGGACTCCA\n",
      "> ID8\n",
      "TGATATGTTGCAGCTTTTCCTGTGGTGTTCTCTGTGA\n",
      "> ID9\n",
      "AAGAGGACCATCCCTAATGCTGATGTGATGCTGA\n",
      "> ID1\n",
      "TGACAGACTAATTGCATCTACGGGGATGAGAATCTGC\n",
      "> ID17\n",
      "TTCTTCTCCCACTGCCCTTTGAAATAGATTGCGTTCA\n",
      "> ID11\n",
      "GTTGGCTTCACTGAGTGAATTCATGGTGAACTCGATG\n",
      "> ID6\n",
      "TATGAGCCAAGAGAATCTGTTGTTGCCAGCAATCAGT\n",
      "> ID7\n",
      "TTTTCTCCGAAGAGCTTGTTGGCGATCTTCAGC\n",
      "> ID13\n",
      "CCATTCCCATGGTTCTCAACGTGTCCTTGAGG\n",
      "> ID15\n",
      "GATCTTTGCCTTTGTATGGTATTTCCAGGACCTTGGC\n"
     ]
    }
   ],
   "source": [
    "to_print = gene_probes_dict[\"NM_006919\"]\n",
    "for i, row in to_print.iterrows():\n",
    "    print(\"> ID\" + str(i))\n",
    "    print(row[\"sequence\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "6d71097d-6e8b-48a9-aaf5-c8fbdf1bec11",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SERPINB3:\n",
    "# 13 (32 nt), 2 (32 nt), 3 (30 nt), 7 (33 nt) are good (but they all have the SERPINB3 vs SERPINB4 problem --> SERPINB4 is slightly expressed) --> show that they essentially share probes and both have no expression (ENCODE shows expression for one exon in one track for SERPIN4 (might be an error))\n",
    "# 12, 5, 9 are bad\n",
    "# at 37, there are a lot that work --> after trying to remove some nt, \n",
    "# 4, 16, 0, 14\n",
    "\n",
    "# all cut to 30 nt, removing from front\n",
    "# 13, 2, 3\n",
    "# 4, 16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07a3c720-841f-4047-bf06-6ec8df0f29c4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
